airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/PostgresTestDatabase.java

```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.testutils;
+
+import com.google.common.collect.ImmutableMap;
+import io.airbyte.cdk.db.Database;
+import io.airbyte.cdk.db.PostgresUtils;
+import io.airbyte.cdk.db.factory.DSLContextFactory;
+import io.airbyte.cdk.db.factory.DatabaseDriver;
+import io.airbyte.cdk.db.jdbc.JdbcUtils;
+import io.airbyte.cdk.integrations.util.HostPortResolver;
+import io.airbyte.commons.io.IOs;
+import io.airbyte.commons.string.Strings;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.jooq.DSLContext;
+import org.jooq.SQLDialect;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.testcontainers.containers.Network;
+import org.testcontainers.containers.PostgreSQLContainer;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.MountableFile;
+
+/**
+ * {@link PostgresTestDatabase} is a convenience object which allows for efficient use of
+ * {@link PostgreSQLContainer} instances in tests. Each test container is shared throughout the
+ * whole JVM. Isolation is performed by creating a new database and a new user for each
+ * {@link PostgresTestDatabase} instance. These are dropped when the instance is closed.
+ */
+public class PostgresTestDatabase implements AutoCloseable {
+
+  /**
+   * Create a new {@link PostgresTestDatabase} instance.
+   *
+   * @param imageName base image to use for the underlying {@link PostgreSQLContainer}.
+   * @param methods {@link ContainerFactory} methods that need to be called.
+   * @return a new {@link PostgresTestDatabase} instance which may reuse a shared
+   *         {@link PostgreSQLContainer}.
+   */
+  static public PostgresTestDatabase make(String imageName, String... methods) {
+    final String imageNamePlusMethods = Stream.concat(
+        Stream.of(imageName),
+        Stream.of(methods))
+        .collect(Collectors.joining("+"));
+    final ContainerFactory factory = ContainerFactory.LAZY.computeIfAbsent(imageNamePlusMethods, ContainerFactory::new);
+    return new PostgresTestDatabase(factory.getOrCreateSharedContainer());
+  }
+
+  private PostgresTestDatabase(PostgreSQLContainer<?> sharedContainer) {
+    this.container = sharedContainer;
+
+    this.suffix = Strings.addRandomSuffix("", "_", 10);
+    try {
+      this.tmpDir = Files.createTempDirectory("dir" + suffix);
+    } catch (final IOException e) {
+      throw new UncheckedIOException(e);
+    }
+    this.tmpDir.toFile().deleteOnExit();
```

Comment:
I don't think it will delete a directory that's not empty but I might be wrong

Reply:
I think you're right.